### PR TITLE
Better waitEvents implementation for *nix-like systems

### LINKED ISSUE #53: Better waitEvents implementation for *nix-like systems
https://github.com/java-native/jssc/blob/0d0e6f7f7210ea8679571c962ea233ae428e8c1f/src/cpp/_nix_based/jssc.cpp#L716

So, this line mentions that there's no analogue for `WaitCommEvent` from WinApi. However, isn't that what `poll`/`epoll`/`select` functions are for? My concern is that for me using this library on Linux causes a memory leak and garbage collection around every 30 seconds (with default flags) when the app is just idle, and I don't think there's a good reason for that.

image

On Windows this behaviour is not present, of course. Now, I have zero experience with C++ code, so I probably shouldn't try to make a PR with critical code, but could someone take a look at it? Here's the [example](https://stackoverflow.com/questions/7035526/read-serial-data-without-high-cpu-use#8832342) on SO of using `poll` on serial port and it's very close to what jSerialComm [does](https://github.com/Fazecast/jSerialComm/blob/53449d43bcbcf3396079752469a0ac276b8f5e83/src/main/c/Posix/SerialPort_Posix.c#L460-L471):
```
JNIEXPORT jint JNICALL Java_com_fazecast_jSerialComm_SerialPort_waitForEvent(JNIEnv *env, jobject obj, jlong serialPortFD)
{
	// Initialize the waiting set
	if (serialPortFD <= 0)
		return 0;
	struct pollfd waitingSet = { serialPortFD, POLLIN, 0 };


	// Wait for a serial port event
	if (poll(&waitingSet, 1, 1000) <= 0)
		return 0;
	return (waitingSet.revents & POLLIN) ? com_fazecast_jSerialComm_SerialPort_LISTENING_EVENT_DATA_AVAILABLE : 0;
}
```
Then there's this [blogpost](https://www.cnblogs.com/darryo/p/selectpollepoll-on-serial-port.html) with three different implementations. I just don't know how to connect all that to JNI and what to do with all the lines statuses.

### PR DESCRIPTION
**Pull request by graynk**
Should close #53

### CODE REVIEW

**Comment by tresf** [comment]
Thanks!

* Please revert the compile lib, all of them need to be rebuilt after this change, so providing one architecture isn't needed.  Furthemore, we build from older kernels so for consistency reasons, we'll do this with the next release.
* I would caution against removing includes like `#include <stdio.h>`.  This is probably safe because `NULL` is imported by `time.h`, but I would recommend reverting that change although it won't affect whether or not we merge it.  I'm just not certain that all old and obscure compilers the project uses are compliant with the C specification so imports should be very carefully removed.

File: src/main/cpp/_nix_based/jssc.cpp [hunk 1]
```diff
@@ -723,6 +723,9 @@ JNIEXPORT jobjectArray JNICALL Java_jssc_SerialNativeInterface_waitEvents
     jclass intClass = env->FindClass("[I");
     jobjectArray returnArray = env->NewObjectArray(sizeof(events)/sizeof(jint), intClass, NULL);
 
+    struct pollfd waitingSet = { static_cast<int>(portHandle), POLLIN, 0 };
+    poll(&waitingSet, 1, 1000);
```
Comment (tresf) [inline src/main/cpp/_nix_based/jssc.cpp:727]: `poll` seem to have [bad portability](https://cr.yp.to/docs/unixport.html) for some unixes, please add a precompiler check to ensure this feature is available.

**Review by tresf — CHANGES_REQUESTED** [review]

File: src/main/java/jssc/SerialPort.java [hunk 2]
```diff
@@ -1301,13 +1301,6 @@ public void run() {
                         }
                     }
                 }
-                //Need to sleep some time
-                try {
-                    Thread.sleep(0, 100);
-                }
-                catch (Exception ex) {
-                    //Do nothing
-                }
```
Comment (tresf) [inline src/main/java/jssc/SerialPort.java:1310]: Per comment above, the likely needs to be conditional based on the availability of `poll()`.  Please adjust accordingly.

**Review by tresf — CHANGES_REQUESTED** [review]

File: src/main/cpp/_nix_based/jssc.cpp (hunk 1 above)
Comment (graynk) [reply src/main/cpp/_nix_based/jssc.cpp]: This page seems incredibly old, it seems poll was added in 2005 to Mac OS (though it was sort of [broken](https://daniel.haxx.se/blog/2016/10/11/poll-on-mac-10-12-is-broken/) for some time). I am not sure how to specifically check for `poll` availability with precompiler checks, so the best I could think of is to check whether we're on Mac

File: src/main/java/jssc/SerialPort.java (hunk 2 above)
Comment (graynk) [reply src/main/java/jssc/SerialPort.java]: Same thing here: I am not sure how to check specifically for `poll` availability, especially from Java, so the best thing I could think of is to check whether we're on Mac OS

**Comment by graynk** [comment]
I put back `#include <stdio.h>` and I've got the libjssc.so from the latest master (couldn't just revert the commit)

File: src/main/cpp/_nix_based/jssc.cpp (hunk 1 above)
Comment (tresf) [reply src/main/cpp/_nix_based/jssc.cpp]: Agreed.  I'd be ok with a 10.9+ requirement, but 10.12+ is probably too new.  We can revisit this solution for Mac at a later time.

File: src/main/java/jssc/SerialPort.java [hunk 3]
```diff
@@ -116,6 +116,8 @@
     private static final int PARAMS_FLAG_PARMRK = 2;
     //<- since 2.6.0
 
+    private static final boolean isMac = System.getProperty("os.name").toLowerCase().contains("mac");
```
Comment (tresf) [inline src/main/java/jssc/SerialPort.java:119]: Please use existing flags, e.g.

```
SerialNativeInterface.getOsType() == SerialNativeInterface.OS_MAC_OS_X)
```

**Comment by tresf** [comment]
> I put back `#include <stdio.h>` and I've got the libjssc.so from the latest master (couldn't just revert the commit)

This still seems to have modified the permissions, which shouldn't be needed.

**Comment by graynk** [comment]
> This still seems to have modified the permissions, which shouldn't be needed.

Sorry about that, fixed.

File: src/main/java/jssc/SerialPort.java (hunk 3 above)
Comment (graynk) [reply src/main/java/jssc/SerialPort.java]: Done

**Comment by tresf** [comment]
@graynk sorry for the delay, it appears this will likely be superseded by #90.  I'd like to invite you to help review the new code.

**Comment by graynk** [comment]
> @graynk sorry for the delay, it appears this will likely be superseded by #90. I'd like to invite you to help review the new code.

No problem. I looked at the code and left a couple of questions. Will close this now.

**Comment by centic9** [comment]
The changes in #90 adjusted "readBytes", but this PR was about adding polling in "waitEvents". 

I hoped to see a reduction of CPU usage of the LinuxEventThread (which consumes up to 10% CPU on a Laptop for me) with this PR, but #90 does not seem to help there. 

Would it be useful to apply the changes to waitEvents in this PR as well so the main event loop polls for events instead of busy-looping and sleeping?

**Comment by GMKennedy** [comment]
> 
> 
> The changes in #90 adjusted "readBytes", but this PR was about adding polling in "waitEvents".
> 
> I hoped to see a reduction of CPU usage of the LinuxEventThread (which consumes up to 10% CPU on a Laptop for me) with this PR, but #90 does not seem to help there.
> 
> Would it be useful to apply the changes to waitEvents in this PR as well so the main event loop polls for events instead of busy-looping and sleeping?

centic9 sorry for the late response on this, had to do quite a bit of background research.  Could you give me a little more information on your test environment and what you are using to profile the cpu spikes you are seeing?  I believe that the issue is resolvable, but the fix involves more than just modifying "waitEvents" and I want to confirm some of my suspicions with testing on my side.  We still appear to be having some issues with opening the serial port in blocking mode

**Comment by GMKennedy** [comment]
@graynk are you seeing performance issues with the most recent release version? How is your test environment configured?

**Comment by graynk** [comment]
> @graynk are you seeing performance issues with the most recent release version? How is your test environment configured?

Sorry, actually I haven't been using this library for a long time now, I've moved on from that project and changed my employer since then.

**Comment by centic9** [comment]
> @centic9 sorry for the late response on this, had to do quite a bit of background research. Could you give us a little more information on your test environment and what you are using to profile the cpu spikes you are seeing? We still appear to be having some issues with opening the serial port in blocking mode (at least this appears to be the case)


No problem, I have the following sample-application and the following impact on CPU (Intel(R) Core(TM) i7-8650U, Linux reports 8 entries in /proc/cpuinfo) when I run this locally:

* Running with "current org.scream3r:jssc:2.8.0": Around 7-8% CPU usage
* Removing the call to device.addEventListener() drops CPU to 0%
* Running with "io.github.java-native:jssc:2.9.3": Similar 7-8% CPU usage
* Running with a locally patched version which adds the poll and also increases the sleep in the loop (see below for actual patch): 0,2% CPU


For sampling CPU I simply use JVisualVM, it shows that the "EventThread" is using up the CPU in the first two cases and "LinuxEventThread.run()" is the method which consumes almost all the CPU.


Sample application:

         public class TM1638TestApp {
                public static void main(String[] args) throws IOException, InterruptedException {
                    System.out.println("Starting Control");
                    try (Control control = new Control()) {
                        System.out.println("Connecting");
                        control.connect("/dev/ttyUSB0");
            
                        for (int i = 0; i < 1000; i++) {
                            Thread.sleep(1000);
                        }
                    }
                }
            
                public static class Control implements AutoCloseable {
                    private SerialPort device;
            
                    public void connect(String comPort) throws IOException {
                        System.out.println("Setting up communication with Arduino device via port " + comPort);
            
                        device = new SerialPort(comPort);
            
                        try {
                            device.openPort();
                            device.setParams(SerialPort.BAUDRATE_19200, SerialPort.DATABITS_8, SerialPort.STOPBITS_1, SerialPort.PARITY_NONE);
            
                            device.addEventListener(new Control.DeviceEventHandler());
                        } catch (SerialPortException e) {
                            close();
            
                            throw new IOException("While opening port " + comPort, e);
                        }
                    }
            
                    @Override
                    public synchronized void close() throws IOException {
                        System.out.println("Shutting down Control");
            
                        if (device != null) {
                            try {
                                if(device.isOpened()) {
                                    System.out.println("Close the device-port");
                                    device.closePort();
                                }
                            } catch (SerialPortException e) {
                                throw new IOException(e);
                            }
                            device = null;
                        }
                    }
            
                    private static class DeviceEventHandler implements SerialPortEventListener {
                        @Override
                        public void serialEvent(SerialPortEvent event) {
                            // process the event
                        }
                    }
                }
            }


Patch to add "poll" and adjust sleep in inner loop:

```
Index: src/main/cpp/_nix_based/jssc.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/cpp/_nix_based/jssc.cpp b/src/main/cpp/_nix_based/jssc.cpp
--- a/src/main/cpp/_nix_based/jssc.cpp	(revision 003dd4f31b749f0c0c151203742c3b207526e15a)
+++ b/src/main/cpp/_nix_based/jssc.cpp	(revision 133da905073a79a62911f6981e010159ea26a164)
@@ -799,10 +799,15 @@
     jclass intClass = env->FindClass("[I");
     jobjectArray returnArray = env->NewObjectArray(sizeof(events)/sizeof(jint), intClass, NULL);
 
+    #ifdef POLLIN
+        struct pollfd waitingSet = { static_cast<int>(portHandle), POLLIN, 0 };
+        poll(&waitingSet, 1, 1000);
+    #endif
+
     /*Input buffer*/
     jint bytesCountIn = 0;
     ioctl(portHandle, FIONREAD, &bytesCountIn);
Index: src/main/java/jssc/SerialPort.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/jssc/SerialPort.java b/src/main/java/jssc/SerialPort.java
--- a/src/main/java/jssc/SerialPort.java	(revision 133da905073a79a62911f6981e010159ea26a164)
+++ b/src/main/java/jssc/SerialPort.java	(revision 5e45c8444de5bd1d465467c04d2bf660f3d15855)
@@ -1382,9 +1382,10 @@
                 try {
-                    Thread.sleep(0, 100);
+                    Thread.sleep(0, 500);
                 }
                 catch (Exception ex) {
                     //Do nothing
```

**Pull request closed without merging**